### ISSUE TITLE
Weights calculated from propensity score network are all clipped to 0.9

### ISSUE THREAD

**Issue by XinranSong** [yccm/DiffPO#1]
I noticed a issue in [main_model.py (line 147)](https://github.com/yccm/DiffPO/blob/main/src/main_model.py#L147). The weights, which should always be greater than 1, are all clipped to 0.9. Won't this alter the intended orthogonal diffusion loss, making it irrelevant to propensity scores for selection bias?

Could you clarify the rationale behind this?

**Comment by XinranSong** [comment]
Thanks for the fix, but this raises a much bigger issue. If the results in your paper were obtained using the original clipped weights, then the propensity score is not being used in the way claimed in the paper. This creates an inconsistency between the method described and what was actually implemented.

Could you please clarify whether the published results were impacted by this issue. If so, addressing this inconsistency and providing a correction would be necessary to ensure the accuracy of the reported results.

In fact, I ran the code without the clip on the weights and with your changes in clipping the propensity score (pi in the code), the RMSE of some datasets of ACIC 2018 goes up to thousands, even tens of thousands.

**Comment by yccm** [comment]
No, the reported results did not include the clipping, that's why I added that clipping propensity score in the code later to ensure other dataset's numerical stability. It is consistent with the paper.

Do you ensure you have properly processed the datasets? I had that issue when the data itself was not properly normalized before training.

**Comment by XinranSong** [comment]
I used the exact same preprocessing Jupyter notebook you provided. The data are normalized. And I see RMSE behaving normal with the 0.9 weights but they would explode without the clipping, and clipping propensity score does not help. For example, with the dataset ACIC 2018, dataset ID = 0c9b0384efcb41348eee39d19e2c285b, the y0 validation RMSE is now 10244 and y1 validation RMSE 7583.

**Comment by yccm** [comment]
That is weird, maybe you should first try running the model without weighting to see if it can still get reasonable results. If not, the hyperparameters need to be tuned first. If without weighting you can have reasonable results. Then the exploding comes from the numerical issue caused by reweighting. I think you can still add the weight clipping if it goes exploding in your case. It is normal to have weights clipped in machine learning tasks, regardless of whether using the propensity score or not. You can set the weight clipping value to a larger range. Clipping the propensity score aims for the same goal. That is to keep a robust training.

**Comment by XinranSong** [comment]
Thank you for your response, but there seems to be a deeper issue here. One of the core contribution of the paper is the proposed reweighting mechanism. However, based on your explanation and the code provided:
	1.	Without the reweighting, doesn’t the method degenerate to normal diffusion models? If so, the implementation fails to reflect the core innovation described in the paper.
	2.	The current demo code, which includes the clipping of propensity scores, does not work as expected. This suggests that the reweighting mechanism (as claimed in the paper) has not been properly implemented or tested.
	3.	If the code only works after removing or incorrectly clipping the reweighting, then the core contribution of the paper becomes invalid, as the proposed innovation is not actually functional.

**Comment by yccm** [comment]
1. I think you didn't get what I meant there. You said you met the problem of exploding. I was helping you to solve that. I don't know why it happens to you. If you have any experience in training deep learning models, you should know there can be various reasons. Checking is just the first step, as no one knows what causes your failed experiments. 

2. In practice, propensity scores can cause unstable training. Even though clipping the propensity scores to a range, it can happen small probability values lead to large weights that can destabilize training. Directly clipping the weights can be a more direct way to prevent any single sample from contributing excessively to the loss. That's why I suggest if you do not have the exploding problem when your normal diffusion model works, you can still add directly clipping the weights to ensure stability. 

3. The paper proposed the method with a theoretical guarantee. That is the main contribution. I am happy you are interested in our work and maybe for future use in your research. If you were asking me why you didn't get the expected results, I am here trying to help you to make your side work.

**Comment by XinranSong** [comment]
Thank you for your response and the efforts in editing your previous responses and deleting your previous "fix". Now the code you left on GitHub still has the mistake I initially pointed out.[main_model.py (line 147)](https://github.com/yccm/DiffPO/blob/main/src/main_model.py#L147)

Now this code works, but it is clipping the weights all to 0.9, making the algorithm a vanilla diffusion model, without the novelty of orthogonal loss. Because the weights = (t_batch / pi_hat[:, 1]) + ((1 - t_batch) / pi_hat[:, 0]) is always bigger than 1. In fact, you made the mistake even in the paper, where you are mapping the function to R+, it should be (1, infinity).

<img width="827" alt="Image" src="https://github.com/user-attachments/assets/59693b9e-b748-45ad-85a2-612f214f4e9a" />

Lastly, I would encourage a more collaborative tone in discussions like this. Acknowledging mistakes—whether in the code, theory, or paper—helps move the field forward and maintains credibility.
